docs/adr/reconstruction.md

```diff
@@ -0,0 +1,286 @@
+- Penalty system for delayed updates
+
+#### 3. Reconstruction State Management
+
+3.1. Global State Components
+```go
+type RemoteState struct {
+    coords [][]peers    // Peer lists by coordinates
+    rows []peers        // Peers with row data
+    cols []peers        // Peers with column data
+    available bitmap    // Available samples bitmap
+}
+
+// Basic peer list structure.Structure might be replaced later to implement better 
+//peer scoring mechanics 
+type peers []peer.ID
+```
+
+3.2. State Query Interface
+
+Remote state:
+- GetPeersWithSample(Coords) -> []peer.ID
+- GetPeersWithAxis(axisIdx, axisType) -> []peer.ID
+- Available() -> bitmap
+
+Local state:
+- InProgress() -> bitmap. Tracks ongoing fetch sessions to prevent requesting duplicates
+- Have() -> bitmap.  Tracks 
+
+
+
+#### 4. Data Request
+There should be global per block coordinator process that will be responsible for managing the data request process.
+
+4.1. Request Initiation may have multiple strategies:
+- Immediate request of all missing samples bitmap receipt
+- Optional: Delayed start for bandwidth optimization
+    - Wait for X% peer responses
+    - Fixed time delay
+    - Complete EDS availability
+    - Pre-confirmation threshold
+    - Combination of conditions
+
+4.1.1 what and were (to request) biggest question.
+
+First iteration of decision engine can be done as simple as possible to allow easier testing of other components and prove the concept. The base properties should be:
+- Eliminate requests for duplicate data
+- Do not request data, that can be derived from other data. Request just enough data for successful reconstruction
+
+
+Following potential optimization trategies
+- Skip encoded derivable data in response by requesting from peers that have all shares from the same rows/columns
+- Optimize proof sizes through range requests
+- Optimize proof sizes through subtree proofs, if adjacent subroots are stored
+- Parallel request distribution to reduce network load on single peers
+- Request from peers with smaller latency
+
+
+4.2. Sample Request Protocol
+
+4.2.1. Request 
+
+- Update InProgress bitmap before request initiation
+- Use shrex for data retrieval. Send bitmap for data request. Shrex has built-in support for requesting rows and samplies, however bitmap-based data retrieval is more general and would be easier to support in future, because it allows requesting multiples pieces of data with single request instead of multiple requests
+```go 
+type SampleRequest struct {
+    Samples *roaring.Bitmap
+}
+```
+
+
+4.2.2. Response 
+- (base version) Server should respond with samples with proofs
+
+Optiomisations:
+- Server may split response into multiple parts
+Each part contains packed samples in Range format.
+- Each part of shoold have specified message prefix. It would allow client to identify sample container type, which would help to maintain backwards compatibility in case of breaking changes in packing algorithm
```

Comment:
```suggestion
- Each part of should have specified message prefix. It would allow client to identify sample container type, which would help to maintain backwards compatibility in case of breaking changes in packing algorithm
```